sdk/identity/azure-identity/azure/identity/aio/_authn_client.py

```diff
@@ -65,30 +65,38 @@ def __init__(
         token = self._deserialize_and_cache_token(response=response, scopes=scopes, request_time=request_time)
         return token
 
-    async def obtain_token_by_refresh_token(self, scopes: "Iterable[str]", username: str) -> "Optional[AccessToken]":
+    async def obtain_token_by_refresh_token(self, scopes: "Iterable[str]", username: "Optional[str]") -> "Optional[AccessToken]":
         """Acquire an access token using a cached refresh token. Returns ``None`` when that fails, or the cache has no
         refresh token. This is only used by SharedTokenCacheCredential and isn't robust enough for anything else."""
 
-        # find account matching username
-        accounts = self._cache.find(TokenCache.CredentialType.ACCOUNT, query={"username": username})
-        for account in accounts:
-            # try each refresh token that might work, return the first access token acquired
-            for token in self.get_refresh_tokens(scopes, account):
-                # currently we only support login.microsoftonline.com, which has an alias login.windows.net
-                # TODO: this must change to support sovereign clouds
-                environment = account.get("environment")
-                if not environment or (environment not in self._auth_url and environment != "login.windows.net"):
-                    continue
+        # if an username is provided, restrect our search to accounts that have that username
+        query = {} if username is None else {"username": username}
+        accounts = self._cache.find(TokenCache.CredentialType.ACCOUNT, query=query)
 
-                request = self.get_refresh_token_grant_request(token, scopes)
-                request_time = int(time.time())
-                response = await self._pipeline.run(request, stream=False)
-                try:
-                    return self._deserialize_and_cache_token(
-                        response=response, scopes=scopes, request_time=request_time
-                    )
-                except ClientAuthenticationError:
-                    continue
+        # Only continue if we found exactly one account. We could find zero if the cache was empty (or there was
+        # no account with the specified username in the cache) and we could find more than one in the case where
+        # username was None and the cache contained multiple identities.
+        if len(accounts) != 1:
+            return None
+
+        account = accounts[0]
```

Comment:
I'm wondering if the intent was to bail if (and only if) user name was not specified AND there is more than one account. The previous version of the code allowed for multiple accounts matching the provided user name and tried them all.

In other words, should the check be:

```python
if username is None and len(accounts) != 1:
    return None
```

and the loop around accounts put back?

Reply:
That was the intent. It's possible for the cache to contain multiple accounts representing the same identity. Lacking any obvious reason to prefer one over another, I chose the first one that works.

Reply:
Thanks for the background here, @chlowell!  Sorry to have changed the behavior here, I'll keep the loop and only require `len(accounts)` to be 1 when `username` is falsy.